pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go

```diff
@@ -214,8 +251,17 @@ func (pl *DefaultPreemption) FindCandidates(ctx context.Context, state *framewor
 		}
 		klog.Infof("from a pool of %d nodes (offset: %d, sample %d nodes: %v), ~%d candidates will be chosen", len(potentialNodes), offset, len(sample), sample, numCandidates)
 	}
-
-	return dryRunPreemption(ctx, pl.fh, state, pod, potentialNodes, pdbs, offset, numCandidates), nil
+	candidates, nodeStatus := dryRunPreemption(ctx, pl.fh, state, pod, potentialNodes, pdbs, offset, numCandidates)
+	// Build the PreemptionFitError only when there is no candidates could be found, and send it back to main scheduler process where it will tell the details on the reason
+	// why the pod is failed to find a node to scheduler even after the pod preemption.
+	if len(candidates) == 0 && len(nodeStatus) != 0 {
+		return candidates, &PreemptionFitError{
```

Comment:
Lets change `FindCandidates` to return three values, candidates, filteredNodesStatuses and error.

Reply:
this is just so external preemption plugins that use this interface can benefit from this info, but this is less critical for this PR.

Reply:
okay, I will take care it after this one.

Reply:
please see: https://github.com/kubernetes/kubernetes/pull/98129